### PR TITLE
Approximated GPU timings in SyncGPU/Single Core

### PR DESCRIPTION
**Pull request by stenzek**
Well, approximated for the vertex stage anyway. The EFB copy/texture preload timings are based on phire's earlier PR.

### CODE REVIEW

File: Source/Core/VideoCommon/VertexLoaderManager.cpp [hunk 1]
```diff
@@ -248,11 +251,42 @@ static VertexLoaderBase* RefreshLoader(int vtx_attr_group, bool preprocess = fal
   return loader;
 }
 
-int RunVertices(int vtx_attr_group, int primitive, int count, DataReader src, bool is_preprocess)
+void UpdateCyclesPerVertex()
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderManager.cpp]: What do you think about caching the cycles_per_vertex as attribute of our vertex loaders?

**Comment by RinMaru** [comment]
will this improve perf on single core i have some games to test if thats the case

**Comment by stenzek** [comment]
@RinMaru not directly. If a game has an uncapped internal framerate, it will cause the framerate to drop, therefore the emulation can run "faster" overall, and require less CPU to do 100%. Same for 30fps games which run at 60fps in Dolphin. But for games which run against vsync (either 30 or 60), it won't make any difference.

**Comment by RinMaru** [comment]
so it will mostly affect games with varieable fps your saying?

File: Source/Core/VideoCommon/OpcodeDecoding.cpp [hunk 2]
```diff
@@ -220,18 +265,22 @@ u8* Run(DataReader src, u32* cycles, bool in_display_list)
...
+              vtx_attr_group, primitive, num_vertices, src, is_preprocess, totalCycles);
+          if (bytes < 0)
+            goto end;
 
-        src.Skip(bytes);
+          src.Skip(bytes);
 
-        // 4 GPU ticks per vertex, 3 CPU ticks per GPU tick
-        totalCycles += num_vertices * 4 * 3 + 6;
+          // TODO: Where are these 6 cycles coming from?
```
Comment (phire) [inline Source/Core/VideoCommon/OpcodeDecoding.cpp]: The pattern throughout this switch block is 6 cycles for the initial cmd byte, and then an additional 6 cycles if the command has an argument.

By  that logic, shouldn't we be adding 12 cycles? 

Then again, I don't know if these timings were guesses, based on some measurement or based on some obscure documentation. Looks like @skidau added these timings.

File: Source/Core/VideoCommon/VertexLoaderManager.cpp [hunk 3]
```diff
@@ -248,11 +251,42 @@ static VertexLoaderBase* RefreshLoader(int vtx_attr_group, bool preprocess = fal
...
 {
-  if (!count)
-    return 0;
+  // Note: These timings do not consider the post-transform vertex cache. These counts are verified
+  // on hardware based on the result of GX_ReadClksPerVtx().
```
Comment (phire) [inline Source/Core/VideoCommon/VertexLoaderManager.cpp]: Were these timings measured with direct or indirect vertex data?

I assume using vertex buffers has some impact on timing. 

Also, would vertices with a large amount of data (direct or indirect) cost more?

File: Source/Core/VideoCommon/VertexLoaderManager.cpp [hunk 4]
```diff
@@ -248,11 +251,42 @@ static VertexLoaderBase* RefreshLoader(int vtx_attr_group, bool preprocess = fal
...
+
+  // Base cost - 4 GPU cycles per vertex.
+  s_cycles_per_vertex = 4;
+
+  // Each enabled color adds a cycle.
+  // Note: This is the colors written to the next stage, not in the vertex format.
+  s_cycles_per_vertex += xfmem.numChan.numColorChans;
+
+  // Each enabled texgen adds 3 cycles.
```
Comment (phire) [inline Source/Core/VideoCommon/VertexLoaderManager.cpp]: one gpu cycle?

**Comment by nbohr1more** [comment]
Rebase?

**Pull request closed without merging**